fix(input): Emit calciteInputInput on clear

macandcheese opened the pull request:
Emits calciteInputInput on clear
Adds tests for clearing on key and clear button press

**Related Issue:**  Resolves #1065 cc @noahmulfinger

## Summary

<!--

Please make sure the PR title and/or commit message adheres to the https://conventionalcommits.org/ specification.

Note: If your PR only has one commit and it is NOT semantic, you will need to either

a. add another commit and wait for the check to update
b. proceed to squash merge, but make sure the commit message is the same as the title.

This is because of the way GitHub handles single-commit squash merges (see https://github.com/zeke/semantic-pull-requests/issues/17)

If this is component-related, please verify that:

- [ ] feature or fix has a corresponding test
- [ ] changes have been tested with demo page in Edge

-->

driskull commented on src/components/calcite-input/calcite-input.tsx:
```diff
@@ -446,6 +450,7 @@ export class CalciteInput {
 
   private clearInputValue = () => {
     this.value = "";
+    this.emitInputInput();
```
could you just `@Watch()` the `value` prop and then emit?

driskull replied:
This is emitted any time the value is cleared or typed. Is that all times `value` is modified? If so, then a `@Watch` makes sense.

driskull reviewed (approved):
Looks good, with suggestion

paulcpederson commented:
Nice, tests as well! :100:

macandcheese commented:
@driskull I updated to emit the event via `@Watch` and added some tests around that for other instances where the value might change, aside from a user explicitly clearing via `Escape` key / clicking "X" button.

driskull reviewed (approved):
👍

driskull commented:
Looks good!

eriklharper commented on src/components/calcite-input/calcite-input.tsx:
```diff
@@ -131,6 +131,13 @@ export class CalciteInput {
   /** input value */
   @Prop({ mutable: true, reflect: true }) value?: string = "";
 
+  @Watch("value") valueWatcher(): void {
+    this.calciteInputInput.emit({
+      element: this.childEl,
+      value: this.value
+    });
+  }
```
Have you checked to make sure this doesn't cause an infinite loop when the `value` prop is being controlled?  It looks like this is the only event that emits the new value, so if the developer is setting the `value` attribute directly as a result of this event being emitted, then the watcher will fire once when `this.value` first changes inside `inputInputHandler` and then when the event is emitted, this might fire again and again each time `value` is set from outside.

driskull replied:
I think it should be ok because the watcher is only called if value is different. right?

eriklharper replied:
Theoretically yes, but would be good to check just to rule out possible race conditions.  The reason I brought this up is that I had to remove an event emitter from a watch method in radio-button because it was causing just this kind of issue.

macandcheese replied:
so would a check look like:

```
@Watch("value") valueWatcher(newValue: string, oldValue: string): void {
  if (newValue !== oldValue) {
    this.calciteInputInput.emit({
      element: this.childEl,
      value: this.value
    });
  }
}

```

eriklharper replied:
Oh, by check I mean try and use calcite-input in a JSX app.  This watcher wouldn't fire unless the new value differs from the old, so this check wouldn't do any good.

macandcheese replied:
Ah, ha. Yeah I thought that was the case with `@Watch`.

eriklharper replied:
Yeah, my hunch/concern is that depending on how fast the user is typing, there could be a sync/timing issue with the value being set internally first and then broadcasted after, which then sets it again from the outside, potentially causing a mismatch.

If there does end up being an issue with this in production apps, we might be able to setup some kind of reusable pattern that we could implement with other similar components where we detect if the `value` attribute is being controlled, which would refrain from storing the value internally and just rely on the event emitter to broadcast the new value, which the app would then set each time it changed from the top down, just as react does with its one-way data flow.

If the value attribute is not controlled, then the current behavior would take place where `this.value` is updated on each change and the event is still emitted.  However, I found for many years that the controlled approach is far superior because it eliminates duplicate local state that can get out of sync with external app state, but it would still make sense to support such a use case since that is how the native input element behaves.
